File: java/countingSort/countingSort.java

class Main {

	private static int[] countingSort(int[] array) {
		int maxElement = Integer.MIN_VALUE;
		for (int e : array) {
			if (e > maxElement) {
				maxElement = e;
			}
		}

		int[] countingArray = new int[maxElement+1];

		for (int e : array) {
			countingArray[e]++;
		}

		for (int i = 1; i < countingArray.length; ++i) {
			countingArray[i] += countingArray[i-1];
		}

		int[] result = new int[array.length];
		for (int i = array.length-1; i >= 0; --i) {
			result[countingArray[array[i]]-1] = array[i];
		}
		return result;
	}

	private static void printArray(int[] array) {
		for (int e : array) {
			System.out.println(e);
		}
	}

	public static void main(String[] args) {
		int[] array = {12, 11, 13, 5, 6};
		int[] result = countingSort(array);
		printArray(result);
	}

}

File: kotlin/countingSort/countingSort.kt

fun countingSort(array: Array<Int>): Array<Int> {
	var max = Int.MIN_VALUE
	for (e in array) {
		if (e > max) {
			max = e
		}
	}

	val countingArray = Array<Int>(max+1){0}
	for (e in array) {
		countingArray[e] ++
	}

	for (i in 1 until countingArray.size) {
		countingArray[i] += countingArray[i-1]
	}

	val result = Array<Int>(array.size){0}
	for (i in (array.size-1) downTo 0) {
		result[countingArray[array[i]]-1] = array[i]
	}
	return result
}

fun printArray(array: Array<Int>) {
	for (e in array) {
		println(e)
	}
}

fun main() {
	val array = arrayOf(12, 11, 13, 5, 6)
	val result = countingSort(array)
	printArray(result)
}

